File: sudoku_solver.js
const fs = require('fs');

class SudokuCSP {
    constructor(boardStr) {
        this.variables = [];
        this.domains = new Map();
        this.neighbors = new Map();

        const rows = boardStr.trim().split(/\r?\n/);
        const board = rows.map(row => row.trim().split('').map(Number));

        for (let r = 0; r < 9; r++) {
            for (let c = 0; c < 9; c++) {
                const varName = `${r},${c}`;
                this.variables.push(varName);
                if (board[r][c] !== 0) {
                    this.domains.set(varName, [board[r][c]]);
                } else {
                    this.domains.set(varName, [1, 2, 3, 4, 5, 6, 7, 8, 9]);
                }
                this.neighbors.set(varName, new Set());
            }
        }

        for (let r = 0; r < 9; r++) {
            for (let c = 0; c < 9; c++) {
                const curr = `${r},${c}`;
                const neighbors = this.neighbors.get(curr);
                
                // Row and Column
                for (let i = 0; i < 9; i++) {
                    if (i !== c) neighbors.add(`${r},${i}`);
                    if (i !== r) neighbors.add(`${i},${c}`);
                }
                
                // Box
                const br = Math.floor(r / 3) * 3;
                const bc = Math.floor(c / 3) * 3;
                for (let i = br; i < br + 3; i++) {
                    for (let j = bc; j < bc + 3; j++) {
                        if (i !== r || j !== c) {
                            neighbors.add(`${i},${j}`);
                        }
                    }
                }
            }
        }
    }

    isConsistent(varName, value, assignment) {
        for (const neighbor of this.neighbors.get(varName)) {
            if (assignment.has(neighbor) && assignment.get(neighbor) === value) {
                return false;
            }
        }
        return true;
    }

    ac3() {
        const queue = [];
        for (const xi of this.variables) {
            for (const xj of this.neighbors.get(xi)) {
                queue.push([xi, xj]);
            }
        }

        while (queue.length > 0) {
            const [xi, xj] = queue.shift();
            if (this.revise(xi, xj)) {
                if (this.domains.get(xi).length === 0) return false;
                for (const xk of this.neighbors.get(xi)) {
                    if (xk !== xj) {
                        queue.push([xk, xi]);
                    }
                }
            }
        }
        return true;
    }

    revise(xi, xj) {
        let revised = false;
        const di = this.domains.get(xi);
        const dj = this.domains.get(xj);
        
        for (let i = di.length - 1; i >= 0; i--) {
            const x = di[i];
            const hasSupport = dj.some(y => x !== y);
            if (!hasSupport) {
                di.splice(i, 1);
                revised = true;
            }
        }
        return revised;
    }
}

function solve(filename) {
    const data = fs.readFileSync(filename, 'utf8');
    const csp = new SudokuCSP(data);

    if (!csp.ac3()) return null;

    const assignment = new Map();
    for (const [varName, domain] of csp.domains) {
        if (domain.length === 1) {
            assignment.set(varName, domain[0]);
        }
    }

    return backtrack(assignment, csp);
}

function backtrack(assignment, csp) {
    if (assignment.size === csp.variables.length) {
        return assignment;
    }

    const unassigned = csp.variables.filter(v => !assignment.has(v));
    // MRV Heuristic
    const varName = unassigned.reduce((min, v) => 
        csp.domains.get(v).length < csp.domains.get(min).length ? v : min, unassigned[0]);

    for (const value of csp.domains.get(varName)) {
        if (csp.isConsistent(varName, value, assignment)) {
            assignment.set(varName, value);
            
            const removals = [];
            let consistent = true;
            
            // Forward Checking
            for (const neighbor of csp.neighbors.get(varName)) {
                if (!assignment.has(neighbor)) {
                    const domain = csp.domains.get(neighbor);
                    const index = domain.indexOf(value);
                    if (index !== -1) {
                        domain.splice(index, 1);
                        removals.push([neighbor, value]);
                        if (domain.length === 0) {
                            consistent = false;
                            break;
                        }
                    }
                }
            }

            if (consistent) {
                const result = backtrack(assignment, csp);
                if (result) return result;
            }

            // Restore
            for (const [n, v] of removals) {
                csp.domains.get(n).push(v);
            }
            assignment.delete(varName);
        }
    }
    return null;
}

function printBoard(assignment) {
    if (!assignment) {
        console.log("No solution found.");
        return;
    }
    for (let r = 0; r < 9; r++) {
        let line = "";
        for (let c = 0; c < 9; c++) {
            line += assignment.get(`${r},${c}`);
        }
        console.log(line);
    }
}

const args = process.argv.slice(2);
if (args.length > 0) {
    const sol = solve(args[0]);
    printBoard(sol);
}

File: sudoku_solver.py
import sys

class SudokuCSP:
    def __init__(self, board_str):
        """
        board_str: A string representing the board (9 lines of 9 digits)
        """
        self.variables = [(r, c) for r in range(9) for c in range(9)]
        self.domains = {}
        self.constraints = {}
        
        # Parse board and initialize domains
        board = [list(line.strip()) for line in board_str.strip().split('\n')]
        for r in range(9):
            for c in range(9):
                val = int(board[r][c])
                if val != 0:
                    self.domains[(r, c)] = [val]
                else:
                    self.domains[(r, c)] = list(range(1, 10))
        
        # Define neighbors for each cell
        self.neighbors = {v: set() for v in self.variables}
        for r, c in self.variables:
            # Row and column neighbors
            for i in range(9):
                if i != c: self.neighbors[(r, c)].add((r, i))
                if i != r: self.neighbors[(r, c)].add((i, c))
            # 3x3 Box neighbors
            br, bc = (r // 3) * 3, (c // 3) * 3
            for i in range(br, br + 3):
                for j in range(bc, bc + 3):
                    if (i, j) != (r, c):
                        self.neighbors[(r, c)].add((i, j))

    def is_consistent(self, var, value, assignment):
        for neighbor in self.neighbors[var]:
            if neighbor in assignment and assignment[neighbor] == value:
                return False
        return True

    def ac3(self):
        queue = []
        for v in self.variables:
            for n in self.neighbors[v]:
                queue.append((v, n))
        
        while queue:
            (xi, xj) = queue.pop(0)
            if self.revise(xi, xj):
                if not self.domains[xi]:
                    return False
                for xk in self.neighbors[xi]:
                    if xk != xj:
                        queue.append((xk, xi))
        return True

    def revise(self, xi, xj):
        revised = False
        for x in self.domains[xi][:]:
            # If no value in Dj allows (x, y) to satisfy constraint
            if not any(x != y for y in self.domains[xj]):
                self.domains[xi].remove(x)
                revised = True
        return revised

def solve_sudoku(filename):
    with open(filename, 'r') as f:
        board_str = f.read()
    
    csp = SudokuCSP(board_str)
    
    # Pre-processing with AC-3
    if not csp.ac3():
        return None
    
    # Backtracking Search
    assignment = {}
    for var, domain in csp.domains.items():
        if len(domain) == 1:
            assignment[var] = domain[0]
            
    solution = backtrack(assignment, csp)
    return solution

def backtrack(assignment, csp):
    if len(assignment) == len(csp.variables):
        return assignment
    
    # MRV Heuristic: Choose variable with minimum remaining values
    unassigned = [v for v in csp.variables if v not in assignment]
    var = min(unassigned, key=lambda v: len(csp.domains[v]))
    
    for value in csp.domains[var]:
        if csp.is_consistent(var, value, assignment):
            assignment[var] = value
            
            # Forward Checking
            removals = []
            consistent = True
            for neighbor in csp.neighbors[var]:
                if neighbor not in assignment:
                    if value in csp.domains[neighbor]:
                        csp.domains[neighbor].remove(value)
                        removals.append((neighbor, value))
                        if not csp.domains[neighbor]:
                            consistent = False
                            break
            
            if consistent:
                result = backtrack(assignment, csp)
                if result:
                    return result
            
            # Backtrack
            for n, v in removals:
                csp.domains[n].append(v)
            del assignment[var]
            
    return None

def print_board(solution):
    if not solution:
        print("No solution found.")
        return
    for r in range(9):
        line = ""
        for c in range(9):
            line += str(solution[(r, c)])
        print(line)

if __name__ == "__main__":
    if len(sys.argv) < 2:
        print("Usage: python sudoku_solver.py <filename>")
    else:
        sol = solve_sudoku(sys.argv[1])
        print_board(sol)
